staging/src/k8s.io/sample-controller/controller.go

```diff
@@ -0,0 +1,378 @@
+	// between multiple control loops
+	informerFactory := informers.NewSharedInformerFactory(clientset, time.Second*30)
+	exampleInformerFactory := intinformers.NewSharedInformerFactory(exampleclientset, time.Second*30)
+
+	return &Controller{
+		clientset:              clientset,
+		exampleclientset:       exampleclientset,
+		informerFactory:        informerFactory,
+		exampleInformerFactory: exampleInformerFactory,
+		deploymentsInformer:    informerFactory.Apps().V1beta1().Deployments(),
+		fooInformer:            exampleInformerFactory.Example().V1alpha1().Foos(),
+		workqueue:              workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Foos"),
+	}, nil
+}
+
+// Run will set up the event handlers for types we are interested in, as well
+// as syncing informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *Controller) Run(stopCh <-chan struct{}) error {
+	glog.Info("Setting up event handlers")
+	// Set up an event handler for when Foo resources change
+	c.fooInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueFoo,
+		UpdateFunc: func(old, new interface{}) {
+			if !reflect.DeepEqual(old, new) {
+				c.enqueueFoo(new)
+			}
+		},
+	})
+	// Set up an event handler for when Deployment resources change. This
+	// handler will lookup the owner of the given Deployment, and if it is
+	// owned by a Foo resource will enqueue that Foo resource for
+	// processing. This way, we don't need to implement custom logic for
+	// handling Deployment resources. More info on this pattern:
+	// https://github.com/kubernetes/community/blob/8cafef897a22026d42f5e5bb3f104febe7e29830/contributors/devel/controllers.md
+	c.deploymentsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.handleObject,
+		UpdateFunc: func(old, new interface{}) {
+			if !reflect.DeepEqual(old, new) {
+				c.handleObject(new)
+			}
+		},
+		DeleteFunc: c.handleObject,
+	})
+
+	// Start the informer factories to begin populating the informer caches
+	glog.Info("Starting informer factories")
+	go c.informerFactory.Start(stopCh)
+	go c.exampleInformerFactory.Start(stopCh)
+
+	// Wait for the caches to be synced before starting workers
+	glog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.fooInformer.Informer().HasSynced, c.deploymentsInformer.Informer().HasSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	glog.Info("Starting workers...")
+	// We use a WaitGroup here so that upon initial shutdown signal, we wait
+	// for the items that are currently being processed to finish processing
+	var wg sync.WaitGroup
+	// Launch two workers to process Foo resources
+	for i := 0; i < 2; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			c.worker()
```

Comment:
the worker call should be inside `wait.Until`, otherwise we won't recover from panic. Compare https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md